### PR TITLE
Add OVH SMS gateway

### PR DESCRIPTION
**Pull request by piti-diablotin**
* Add the ovh api to be able to use the gateway sms with ovh [(website)](https://www.ovhtelecom.fr/sms/)
* Documentation added of OVH in the Administration file.
* It has been tested with the test function and on my testing setting (NC16)

PS:Sorry for the different names in the commits I did not take care of it.

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Thanks a lot for this :pray: 

Unfortunately that `ovh/ovh` package pulls in Guzzle. This will cause conflicts with server and other apps. The upstream package has to sort out https://github.com/ovh/php-ovh/issues/34, then we can integrate `ovh/ovh` in Nextcloud 19 in combination with https://packagist.org/packages/christophwurst/nextcloud-http-client.

**Comment by ChristophWurst** [comment]
Let's see if this can be driven forward https://github.com/ovh/php-ovh/pull/41#issuecomment-576351971 :)

**Comment by ChristophWurst** [comment]
Code looks good btw :)

**Comment by ChristophWurst** [comment]
Sure! That is also what we use for most other gateway integrations :)

We can also migrate to the updated ovh package later :wink:

**Comment by piti-diablotin** [comment]
I have update the Provider with the IClientService from NC.
I have tested with test function provided by the app and it seems working.
The `ovh/ovh Api` are commented but still working.

Would it be possible to merge it now ?

**Comment by piti-diablotin** [comment]
I'm sorry I can't get the DCO despite the `git commit -s` !!! What am I doing wrong ?

File: lib/Service/Gateway/SMS/Provider/Ovh.php [hunk 1]
```diff
@@ -54,20 +80,37 @@ public function __construct(IClientService $clientService,
...
+    $this->attrs['AS'] = $config->getApplicationSecret();
+    $this->attrs['CK'] = $config->getConsumerKey();
+    if (!isset($this->endpoints[$endpoint]))
+			throw new InvalidSmsProviderException("Endpoint $endpoint not found");
+    $this->attrs['endpoint'] = $this->endpoints[$endpoint];
 
-		$smsAccount = $config->getAccount();
+    $this->getTimeDeta();
 
+    /* Use with \Ovh\Api
```
Comment (ChristophWurst) [inline lib/Service/Gateway/SMS/Provider/Ovh.php]: I'm not a fan of :skull:  code. Could you remove it? :)

File: lib/Service/Gateway/SMS/Provider/Ovh.php [hunk 2]
```diff
@@ -54,20 +80,37 @@ public function __construct(IClientService $clientService,
...
+
+    /* Use with \Ovh\Api
 		$smsServices = $conn->get('/sms');
+     */
+
+    $header = $this->getHeader('GET',$this->attrs['endpoint'].'/sms');
+    $response = $this->client->get($this->attrs['endpoint'].'/sms',[
+      'headers' => $header,
+    ]);
+    $smsServices = json_decode($reponse->getBody());
```
Comment (ChristophWurst) [inline lib/Service/Gateway/SMS/Provider/Ovh.php]: your indentation is off. See the diff here on github :)

File: lib/Service/Gateway/SMS/Provider/Ovh.php [hunk 3]
```diff
@@ -78,7 +121,7 @@ public function send(string $identifier, string $message) {
 		if ($smsAccountFound === false) {
 			throw new InvalidSmsProviderException("SMS account $smsAccount not found");
 		}
-		$content = (object) array(
+		$content = array(
```
Comment (ChristophWurst) [inline lib/Service/Gateway/SMS/Provider/Ovh.php]: please use the `[]` array syntax

File: lib/Service/Gateway/SMS/Provider/Ovh.php [hunk 4]
```diff
@@ -100,4 +154,44 @@ public function send(string $identifier, string $message) {
...
+   * @param string $body JSON encoded body content for the POST request
+   */
+  private function getHeader($method,$query,$body='') {
+    $timestamp = time() + $this->attrs['timedelta'];
+    $prehash = $this->attr['AS'].'+'.$this->attr['CK'].'+'.$method.'+'.$query.'+'.$body.'+'.$timestamp;
+    $header = [
+      'Content-Type'      => 'application/json; charset=utf-8',
+      'X-Ovh-Application' => $this->attrs['AK'],
+      'X-Ovh-Timestamp'   => $timestamp,
+      'X-Ovh-Signature'   => '$1$' + sha1($prehash),
```
Comment (ChristophWurst) [inline lib/Service/Gateway/SMS/Provider/Ovh.php]: `sha1` returns a string, hence the `+` is very suspicious.

**Comment by ChristophWurst** [comment]
> I'm sorry I can't get the DCO despite the `git commit -s` !!! What am I doing wrong ?

The sign-off has to be present on all of your commits. I would recommend to [use fixup's](https://dev.to/tmr232/dont-amend-fix) during development and do a final squash before the merge so it's just one commit in the end.

**Comment by piti-diablotin** [comment]
I squashed everything. But Travis fails :(
At least DCO seems fine.
Why would I have `App "Two-Factor Gateway" cannot be installed because it is not compatible with this version of the server.` whereas I did not changed anything in the setting of the app ... ? Did I messed up something ?

**Comment by ChristophWurst** [comment]
We have to fix CI as the nextcloud server versions changed recently. Thus the incompatibility failure.

It's as simple as this: https://github.com/nextcloud/mail/pull/2523

**Comment by ChristophWurst** [comment]
> [What happened to this file ?!](/nextcloud/twofactor_gateway/pull/316/commits/10c002a96919aba4d74eb681642c3b2e6b9fb6c7)

LOL at this clickbait title. Made me want to know what happened :)

**Comment by piti-diablotin** [comment]
How should I make tests to improve code coverage ?

**Comment by piti-diablotin** [comment]
Hi,
Could we proceed ?

**Comment by ChristophWurst** [comment]
> How should I make tests to improve code coverage ?

You can see https://github.com/nextcloud/twofactor_gateway/tree/master/tests for reference. But generally this app doesn't have many tests, unfortunately.

**Pull request merged**